bin/oe/rpc.rs

```diff
@@ -189,44 +244,84 @@ pub fn new_ws<D: rpc_apis::Dependencies>(
-        handler,
-        allowed_origins,
-        allowed_hosts,
-        conf.max_connections,
-        rpc::WsExtractor::new(path.clone()),
-        rpc::WsExtractor::new(path.clone()),
-        rpc::WsStats::new(deps.stats.clone()),
-        conf.max_payload,
-    );
-
-    //	match start_result {
-    //		Ok(server) => Ok(Some(server)),
-    //		Err(rpc::ws::Error::Io(rpc::ws::ErrorKind::Io(ref err), _)) if err.kind() == io::ErrorKind::AddrInUse => Err(
-    //			format!("WebSockets address {} is already in use, make sure that another instance of an Ethereum client is not running or change the address using the --ws-port and --ws-interface options.", url)
-    //		),
-    //		Err(e) => Err(format!("WebSockets error: {:?}", e)),
-    //	}
-    match start_result {
-		Ok(server) => Ok(Some(server)),
-		Err(rpc::ws::Error::WsError(ws::Error {
-			                            kind: ws::ErrorKind::Io(ref err), ..
-		                            })) if err.kind() == io::ErrorKind::AddrInUse => Err(
-			format!("WebSockets address {} is already in use, make sure that another instance of an Ethereum client is not running or change the address using the --ws-port and --ws-interface options.", url)
-		),
-		Err(e) => Err(format!("WebSockets error: {:?}", e)),
-	}
+
+    let mut servers = Vec::default();
+
+    servers.push(format_ws_error(
+        &url,
+        rpc_servers::start_ws(
+            &addr,
+            handler,
+            allowed_origins.clone(),
+            allowed_hosts.clone(),
+            conf.max_connections,
+            rpc::WsExtractor::new(path.clone()),
+            rpc::WsExtractor::new(path.clone()),
+            rpc::WsStats::new(deps.stats.clone()),
+            conf.max_payload,
+        ),
+    )?);
+
+    for endpoint in conf.additional_endpoints {
+        let url = format!("{}:{}", endpoint.interface, endpoint.port);
+        let addr = url
+            .parse()
+            .map_err(|_| format!("Invalid WebSockets listen host/port given: {}", url))?;
+
+        let full_handler = setup_apis(rpc_apis::ApiSet::All, deps);
+        let handler = {
+            let mut handler = MetaIoHandler::with_middleware((
+                rpc::WsDispatcher::new(full_handler),
+                Middleware::new(deps.stats.clone(), deps.apis.activity_notifier()),
+            ));
+            let apis = endpoint.apis.list_apis();
+            deps.apis.extend_with_set(&mut handler, &apis);
+
+            handler
+        };
+
+        servers.push(format_ws_error(
+            &url,
+            rpc_servers::start_ws(
+                &addr,
+                handler,
+                allowed_origins.clone(),
+                allowed_hosts.clone(),
+                conf.max_connections,
+                rpc::WsExtractor::new(path.clone()),
+                rpc::WsExtractor::new(path.clone()),
+                rpc::WsStats::new(deps.stats.clone()),
+                conf.max_payload,
+            ),
+        )?);
+    }
+
```

Comment:
Maybe it is possible to eliminate code duplication in lines 219-234;250-263 and 266-297; and 328-332;340-355 and 358-380